sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,330 @@ import org.apache.spark.util.Utils
+        }
+      }
 
-      case (VALUE_TRUE, BooleanType) =>
-        true
+    case dt: DecimalType =>
+      (parser: JsonParser) => handleNullAndFieldTokens(parser) {
+        parser.getCurrentToken match {
+          case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT) =>
+            Decimal(parser.getDecimalValue, dt.precision, dt.scale)
 
-      case (VALUE_FALSE, BooleanType) =>
-        false
+          case _ => failedConversion(parser, dt)
+        }
+      }
 
-      case (START_OBJECT, st: StructType) =>
-        convertObject(factory, parser, st)
+    case st: StructType =>
+      val fieldConverters = st.map(_.dataType).map(makeConverter)
+      (parser: JsonParser) => handleNullAndFieldTokens(parser) {
+        parser.getCurrentToken match {
+          case START_OBJECT => convertObject(parser, st, fieldConverters)
+          case _ => failedConversion(parser, st)
+        }
+      }
 
-      case (START_ARRAY, ArrayType(st, _)) =>
-        convertArray(factory, parser, st)
+    case at: ArrayType =>
+      val elementConverter = makeConverter(at.elementType)
+      (parser: JsonParser) => handleNullAndFieldTokens(parser) {
+        parser.getCurrentToken match {
+          case START_ARRAY => convertArray(parser, elementConverter)
+          case _ => failedConversion(parser, at)
+        }
+      }
 
-      case (START_OBJECT, MapType(StringType, kt, _)) =>
-        convertMap(factory, parser, kt)
+    case mt: MapType =>
+      val valueConverter = makeConverter(mt.valueType)
+      (parser: JsonParser) => handleNullAndFieldTokens(parser) {
+        parser.getCurrentToken match {
+          case START_OBJECT => convertMap(parser, valueConverter)
+          case _ => failedConversion(parser, mt)
+        }
+      }
 
-      case (_, udt: UserDefinedType[_]) =>
-        convertField(factory, parser, udt.sqlType)
+    case udt: UserDefinedType[_] =>
+      makeConverter(udt.sqlType)
 
-      case (token, dataType) =>
-        // We cannot parse this token based on the given data type. So, we throw a
-        // SparkSQLJsonProcessingException and this exception will be caught by
-        // parseJson method.
-        throw new SparkSQLJsonProcessingException(
-          s"Failed to parse a value for data type $dataType (current token: $token).")
+    case _ =>
+      (parser: JsonParser) =>
+        failedConversion(parser, dataType)
+  }
+
+  /**
+   * This handles nulls ahead before trying to check the tokens, and applies
+   * the conversion function. Also, it checks `FIELD_NAME` if exists and then skip.
+   * If this is called after `START_OBJECT`, then, the next token can be
+   * `FIELD_NAME`. Since the names are kept in `JacksonParser.convertObject`, this `FIELD_NAME`
+   * token can be skipped as below. When this is called after `START_ARRAY`, the tokens become
+   * ones about values until `END_ARRAY`. In this case, we don't have to skip.
+   */
+  private def handleNullAndFieldTokens (parser: JsonParser)(f: => Any): Any = {
```

Comment:
@cloud-fan I literally tried to come up with a good name for about 30 min but I couldn't. So, I just chose a verbose one..